### PR TITLE
aws_hash_combine()

### PR DESCRIPTION
**Pull request by graebm**
We have hashes for single items.
This lets us combine them.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/hash_table_test.c [hunk 1]
```diff
@@ -1258,3 +1258,39 @@ static int s_test_hash_table_byte_cursor_create_find_fn(struct aws_allocator *al
...
+        ASSERT_TRUE(c_bytes[i] != 0);
+    }
+
+    /* Check even low-entropy numbers result in something with high entropy */
+    c = aws_hash_combine(0, 1);
+    for (size_t i = 0; i < sizeof(c); ++i) {
+        ASSERT_TRUE(c_bytes[i] != 0);
+    }
+
+    /* Hash should NOT be cummutative */
```
Comment (danielsn) [inline tests/hash_table_test.c:1292]: commutative

File: tests/hash_table_test.c [hunk 2]
```diff
@@ -1258,3 +1258,39 @@ static int s_test_hash_table_byte_cursor_create_find_fn(struct aws_allocator *al
...
+    /* Using all bytes, right? */
+    uint8_t *c_bytes = (uint8_t *)&c;
+    for (size_t i = 0; i < sizeof(c); ++i) {
+        ASSERT_TRUE(c_bytes[i] != 0);
+    }
+
+    /* Check even low-entropy numbers result in something with high entropy */
+    c = aws_hash_combine(0, 1);
+    for (size_t i = 0; i < sizeof(c); ++i) {
+        ASSERT_TRUE(c_bytes[i] != 0);
```
Comment (danielsn) [inline tests/hash_table_test.c:1289]: Can we do better than that? If we want entropy, we could actually measure entropy.

Comment (graebm) [reply tests/hash_table_test.c:1289]: Yes we could do better, but this is not the test were we would do it.

We use `lookup3.inl`, an existing collection of functions for hashing. We've assumed so far that these do what they say. I just wanted to make sure I didn't do something dumb where the result never had its high or low bytes set.

File: tests/hash_table_test.c (hunk 1 above)
Comment (graebm) [reply tests/hash_table_test.c:1292]: derp. thanks

File: tests/hash_table_test.c (hunk 2 above)
Comment (graebm) [reply tests/hash_table_test.c:1289]: I just took this part of the test out. It was a naive test the quality of the underlying hash function, and this isn't the place for it

File: tests/hash_table_test.c [hunk 3]
```diff
@@ -1258,3 +1258,39 @@ static int s_test_hash_table_byte_cursor_create_find_fn(struct aws_allocator *al
...
+    /* Same inputs gets same results, right? */
+    ASSERT_UINT_EQUALS(c, aws_hash_combine(a, b));
+
+    /* Using all bytes, right? */
+    uint8_t *c_bytes = (uint8_t *)&c;
+    for (size_t i = 0; i < sizeof(c); ++i) {
+        ASSERT_TRUE(c_bytes[i] != 0);
+    }
+
+    /* Check even low-entropy numbers result in something with high entropy */
```
Comment (justinboswell) [inline tests/hash_table_test.c]: If we wanted to test this, wouldn't we just fuzz it?

Comment (graebm) [reply tests/hash_table_test.c]: I just took this part of the test out. This is not the place to test the quality of the underlying hash library

File: source/hash_table.c [hunk 4]
```diff
@@ -991,6 +991,14 @@ uint64_t aws_hash_ptr(const void *item) {
     return ((uint64_t)b << 32) | c;
 }
 
+uint64_t aws_hash_combine(uint64_t item1, uint64_t item2) {
+    uint32_t b = (uint32_t)item2;
```
Comment (feliperodri) [inline source/hash_table.c:995]: This conversion could overflow, but it's intended, right? Also, would it make sense to replace it with the following line? 
```c
uint32_t b = item2 & 0xFFFFFFFF; /* LSB */
uint32_t c = item2 >> 32; /* MSB */
```

Comment (graebm) [reply source/hash_table.c:995]: ok

**Pull request merged**